lib/Platform/Intl/BCP47Parser.cpp

```diff
@@ -0,0 +1,539 @@
+}
+bool isASCIIDigit(char16_t c) {
+  return (c >= '0' && c <= '9');
+}
+bool isASCIILetterOrDigit(char16_t c) {
+  return isASCIILetter(c) || isASCIIDigit(c);
+}
+void toASCIILowerCase(std::u16string &str) {
+  for (size_t i = 0; i < str.length(); i++) {
+    if (str[i] <= 'Z' && str[i] >= 'A') {
+      str[i] -= 'Z' - 'z';
+    }
+  }
+  return;
+}
+void toASCIIUpperCase(std::u16string &str) {
+  for (size_t i = 0; i < str.length(); i++) {
+    if (str[i] <= 'z' && str[i] >= 'a') {
+      str[i] -= 'z' - 'Z';
+    }
+  }
+  return;
+}
+void toASCIITitleCase(std::u16string &str) {
+  for (size_t i = 0; i < str.length(); i++) {
+    if (i == 0 && str[i] <= 'z' && str[i] >= 'a') {
+      str[i] -= 'z' - 'Z';
+    } else if (str[i] <= 'Z' && str[i] >= 'A') {
+      str[i] -= 'Z' - 'z';
+    }
+  }
+  return;
+}
+bool isSubtagSeparator(char16_t c) {
+  return c == u'-';
+}
+bool isCharType(const llvh::ArrayRef<char16_t> str, int min, int max, bool(*charType)(char16_t)) {
+  int length = str.size();
+  if (length == 0) {
+    return false;
+  }
+  
+  if (length < min || length > max) {
+    return false;
+  }
+  
+  for (char16_t c : str) {
+    if (!charType(c)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// tag type functions
+bool isUnicodeLanguageSubtag(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = alpha{2,3} | alpha{5,8};
+  // root case?
+  return isCharType(subtagRef, 2, 3, &isASCIILetter) ||
+    isCharType(subtagRef, 5, 8, &isASCIILetter);
+}
+bool isUnicodeScriptSubtag(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = alpha{4};
+  return isCharType(subtagRef, 4, 4, &isASCIILetter);
+}
+bool isUnicodeRegionSubtag(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = (alpha{2} | digit{3});
+  return isCharType(subtagRef, 2, 2, &isASCIILetter) ||
+    isCharType(subtagRef, 3, 3, &isASCIIDigit);
+}
+bool isUnicodeVariantSubtag(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = (alphanum{5,8} | digit alphanum{3});
+  return isCharType(subtagRef, 5, 8, &isASCIILetterOrDigit) ||
+    isCharType(subtagRef, 3, 3, &isASCIILetterOrDigit);
+}
+bool isUnicodeExtensionAttribute(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = alphanum{3,8};
+  return isCharType(subtagRef, 3, 8, &isASCIILetterOrDigit);
+}
+bool isUnicodeExtensionKey(const llvh::ArrayRef<char16_t> subtagRef) {
+  // https://unicode.org/reports/tr35/#Unicode_Language_and_Locale_Identifiers
+  // = alphanum alpha;
+  if (subtagRef.size() != 2) {
+    return false;
+  }
+  return isASCIILetterOrDigit(subtagRef.front()) && isASCIILetter(subtagRef.back());
```

Comment:
consistency nit (same below)

```suggestion
  return subtagRef.size() == 2 && isASCIILetterOrDigit(subtagRef.front()) && isASCIILetter(subtagRef.back());
```